### ISSUE TITLE
[RFE] Create client library to handle common updates workflow

### ISSUE THREAD

**Issue by joaquimrocha** [kinvolk/nebraska#419]
## Current situation

Nebraska implements the Omaha protocol and thus any clients should be able to interact with it (i.e. perform update requests, and communicate the state back throughout the update workflow) using it.

## Impact

This however requires knowledge of the Omaha protocol itself (what an update request gets as response, what state to send when an update is finished, etc.) and that can become a complex task in itself.

## Ideal future situation

We want to lower the barrier for projects to start using Omaha-managed updates, and thus would like to have a library that requires the minimum configuration/code from clients in order to manage updates for a certain product.

For example, at a minimum, users should have to tell this library what the Nebraska/Omaha endpoint is + what to do with a downloaded update payload. All state communication/handling should be done with default settings by the library.

## **Implementation options

We can start by making it a go package `nebraskaClient github.com/nebraska/client/v1`.

```go
type updater struct {
  nebraskaClient.ContainerImageUpdater
}

func (* updater) Deploy(manifest *nebraskaClient.UpdateManifest) nebraskaClient.Error {
  // Do whatever and then docker run again...
}

// Client initialization
ne := nebraskaClient.NewUpdateEndpoint("https://myserver/app/io.org.App/")
ne.SetUpdater(updater)
ne.SetCurrentVersion("v0.1.2")
ne.SetChannel("beta")

ne.Start()
```

Maybe my go code above is not the best but hopefully gives the idea.

After `ne.Start()`, the Nebraska client starts performing update requests to the server. When an update exists, Nebraska responds with a URL and a metadata. Since we are composing a "container image updater", it's default "Download" implementation will pull + verify the new image, and then our implementation will stop a service and run the image again (we get details from the image in the UpdateManifest struct which is TBD).

## Additional information

This was a rough example, maybe other basic functions related to authentication need to be done. Take it with a grain of sault.

**Comment by vbatts** [comment]
sounds great.

the requirements/outcome:
- anything considered (like openapi/swagger) does not break the existing API used by all the Flatcar machines in the world
- it's very intuitive for users to read and discover how to write code to either Consume from a channel, or publish new content to a channel, using the API and libraries
- it's easy to manage for the project (write once, and generate; as opposed to having to make changes across code and docs manually)

**Comment by joaquimrocha** [comment]
I agree. Please also note that in my description above, it shouldn't affect the server code at all. I am currently writing another issue regarding how to cover different payload response types, and that's where we should be more careful.
For the admin/write CLI we can manage it in the #357 .
